File: apps/api/src/routes/files.ts
import { Hono } from 'hono'
import path from 'node:path'
import { storageService } from '../services/storage.service.js'
import { sanitizeFileName } from '../lib/sanitize.js'

const app = new Hono()

const ALLOWED_PREFIXES = ['generated/', 'uploads/']

app.get('/files/*', async (c) => {
  const raw = c.req.path.replace('/api/files/', '')
  const key = decodeURIComponent(raw)
  const normalized = path.normalize(key)
  if (!key || key.includes('\0') || normalized.startsWith('..') || normalized !== key) {
    return c.json({ error: 'Invalid file path' }, 400)
  }
  if (!ALLOWED_PREFIXES.some((p) => key.startsWith(p))) {
    return c.json({ error: 'Invalid file path' }, 400)
  }

  try {
    const stream = await storageService.download(key)
    if (!stream) {
      return c.json({ error: 'File not found' }, 404)
    }

    const filename = sanitizeFileName(key.split('/').pop() ?? 'file')
    return new Response(stream as ReadableStream, {
      headers: {
        'Content-Disposition': `attachment; filename="${filename}"`,
        'Content-Type': 'application/octet-stream',
      },
    })
  } catch {
    return c.json({ error: 'File not found' }, 404)
  }
})

export default app

File: backend/src/clawbuddy/routers/files.py
"""File serving routes.

Replaces: apps/api/src/routes/files.ts
"""

from __future__ import annotations

import posixpath

from fastapi import APIRouter, Request
from fastapi.responses import Response

from clawbuddy.lib.responses import fail
from clawbuddy.lib.sanitize import sanitize_file_name
from clawbuddy.services.storage import storage_service

router = APIRouter(tags=["Files"])

_ALLOWED_PREFIXES = ("generated/", "uploads/")


@router.get("/files/{file_path:path}")
async def get_file(file_path: str, request: Request) -> Response:
    """Serve a file from object storage."""
    key = file_path
    normalized = posixpath.normpath(key)

    if (
        not key
        or "\0" in key
        or normalized.startswith("..")
        or normalized != key
    ):
        return fail("Invalid file path", status_code=400)  # type: ignore[return-value]

    if not any(key.startswith(p) for p in _ALLOWED_PREFIXES):
        return fail("Invalid file path", status_code=400)  # type: ignore[return-value]

    body = await storage_service.download(key)
    if body is None:
        return fail("File not found", status_code=404)  # type: ignore[return-value]

    filename = sanitize_file_name(key.rsplit("/", 1)[-1] if "/" in key else "file")
    return Response(
        content=body,
        media_type="application/octet-stream",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )
